### PR TITLE
Add option to use steps instead of epochs

### PR DESCRIPTION
**Pull request by ksikka**
~~Status: unit tests pass, but I have not done any manual testing.~~
~~I am hoping we can use the pipeline to test, I will also think of how to add unittest coverage for some of these.~~

Update(2/9): added tests. This is a new kind of test that trains a model and then asserts things about the training process using the logs. We test that it trains to the correct number of steps/epochs, unfreezes backbone, anneals LR according to milestones.

### CODE REVIEW

**Comment by ksikka** [comment]
@themattinthehatt ready for review. A few callouts:
* backbone unfreezing with steps is faster since the warm up ratio is applied to learning rate per step not per epoch. The warm up ratio is currently not configurable.
* I found an issue with limit_train_batches not doing what we think. see `calculate_train_batches`

Some misc changes in this PR:
* Fix bug in model.py: output_dir argument was not converted to Path, causing methods to break with strings
* train now accepts a model_dir argument, user no longer needs to chdir (note-to-self, update docs in this PR)
* train also allows you to skip evaluation, I added this for tests.
* If there's not enough data, compute_metrics will no longer fail when attempting to compute PCA (for singleview)

**Review by themattinthehatt — CHANGES_REQUESTED** [review]
need to figure out limit_train_batches issue before merging

**Comment by ksikka** [comment]
> need to figure out limit_train_batches issue before merging

Fixed the issue. Previously I was trying to use `calculate_train_batches` as the number of steps per epoch. This returned a value of `10` for small train_frames. This is a limit, and it only applies for unsupervised learning. Using it for supervised doesn't make sense.

I've replaced it with a calculation of steps per epoch. See the most recent commit for these code changes.

Additional cleanups in the last commit: 
- Remove the need to call .setup() on a datamodule. the logic now is entirely triggered from calling the constructor.


One caveat with calculate_steps_per_epoch: we cannot precisely calculate steps per epoch in the rare case of small train_frames, and very short training video (small enough that the 10 step minimum would trigger). our calculation would return 10 steps per epoch, when reality is less. This is rare enough that we can ignore it for now.

File: tests/utils/test_scripts.py [hunk 1]
```diff
@@ -72,59 +97,48 @@ def _unsupervised_multi_gpu_cfg(cfg):
...
-def test_get_data_module_num_gpus_0(cfg):
+def test_get_data_module_num_gpus_0(cfg, mocker):
     cfg = _supervised_multi_gpu_cfg(cfg)
     # when num_gpus is set to 0, i.e. from a deprecated config
     cfg.training.num_gpus = 0
-    data_module = get_data_module(cfg, Mock(spec=BaseTrackingDataset))
+    mock_data_module_init = mocker.patch.object(BaseDataModule, '__init__', return_value=None)
+    get_data_module(cfg, Mock(spec=BaseTrackingDataset))
+
+    get_data_module(cfg, Mock(spec=BaseTrackingDataset))
```
Comment (themattinthehatt) [inline tests/utils/test_scripts.py:107]: Can remove both `get_data_module` calls, they don't appear to be used.

Can you say more about the motivation to use this mocker pattern? Why not just create a data module?

File: tests/utils/test_scripts.py [hunk 2]
```diff
@@ -72,59 +97,48 @@ def _unsupervised_multi_gpu_cfg(cfg):
...
+    assert mock_data_module_init.call_args.kwargs["val_batch_size"] == cfg.training.val_batch_size
+    assert mock_data_module_init.call_args.kwargs["test_batch_size"] == cfg.training.test_batch_size
 
 
-def test_get_data_module_multi_gpu_batch_size_adjustment_supervised(cfg):
+def test_get_data_module_multi_gpu_batch_size_adjustment_supervised(cfg, mocker):
     cfg = _supervised_multi_gpu_cfg(cfg)
-    data_module = get_data_module(cfg, Mock(spec=BaseTrackingDataset))
+    mock_data_module_init = mocker.patch.object(BaseDataModule, '__init__', return_value=None)
+    get_data_module(cfg, Mock(spec=BaseTrackingDataset))
```
Comment (themattinthehatt) [inline tests/utils/test_scripts.py:120]: same question - unused?

File: tests/utils/test_scripts.py [hunk 3]
```diff
@@ -72,59 +97,48 @@ def _unsupervised_multi_gpu_cfg(cfg):
...
 
 
 def test_get_data_module_multi_gpu_batch_size_adjustment_unsupervised(
-    cfg, heatmap_dataset, toy_data_dir
+    cfg, heatmap_dataset, toy_data_dir, mocker
 ):
     cfg = _unsupervised_multi_gpu_cfg(cfg)
-    data_module = get_data_module(
+    mock_data_module_init = mocker.patch.object(UnlabeledDataModule, '__init__', return_value=None)
+    get_data_module(
```
Comment (themattinthehatt) [inline tests/utils/test_scripts.py:132]: unused?

File: tests/tblogs/test_lr.py
Comment (themattinthehatt) [inline tests/tblogs/test_lr.py:1]: rename to `tests/test_training/test_lr.py`

File: tests/tblogs/test_train_duration.py
Comment (themattinthehatt) [inline tests/tblogs/test_train_duration.py:1]: rename to `tests/test_training/test_train_duration.py`

File: setup.py [hunk 4]
```diff
@@ -75,7 +75,8 @@ def get_cuda_version():
     "scikit-learn>=1.5.0",
     "seaborn",
     "streamlit",
-    "tensorboard",
+    "tensorboard<3.0.0",
+    "tbparse<1.0.0",
```
Comment (themattinthehatt) [inline setup.py:79]: move to extras_require/dev below

**Review by themattinthehatt — CHANGES_REQUESTED** [review]

File: tests/utils/test_scripts.py (hunk 1 above)
Comment (ksikka) [reply tests/utils/test_scripts.py:107]: One is duplicative, removed.

The get_data_module has the side effect of calling the BaseDataModule constructor, which we then test assertions about.  But we don't need the returned object for this, we use the mocker object to assert what it was called with.

Reason for mocker: Previous code was running the real constructor, with a mock dataset as an input argument. This led to errors once the constructor started calling the _setup method. Mock dataset returns None for all attributes. The specific failure was about dataset.__len__ when the datamodule tries to split the dataset by length.

What the mocker does: replace data module constructor with a fake that does nothing. It just intercepts the args and allows you to assert things about them. This is the ideal thing for a "unit test". It tests just the logic in scripts.py, and exercises none of the code in datamodule.

File: tests/utils/test_scripts.py (hunk 2 above)
Comment (ksikka) [reply tests/utils/test_scripts.py:120]: Side effect of calling data module constructor

File: setup.py (hunk 4 above)
Comment (ksikka) [reply setup.py:79]: oops yeah, done.

File: tests/tblogs/test_train_duration.py
Comment (ksikka) [reply tests/tblogs/test_train_duration.py:1]: done

File: tests/tblogs/test_lr.py
Comment (ksikka) [reply tests/tblogs/test_lr.py:1]: done

File: tests/utils/test_scripts.py (hunk 3 above)
Comment (ksikka) [reply tests/utils/test_scripts.py:130]: side effect of calling constructor
